### PR TITLE
fix: ZodEffect in discriminatedUnion

### PR DESCRIPTION
**Pull request by cdaringe**
# Problem

`.refine` and `.superRefine`d schemas cannot be used in `discriminatedUnion`, but should be able to be.

See https://github.com/colinhacks/zod/issues/2440

# Discussion

The fix is very simple.

1. Support `ZodEffect`s where in the input type is also a `ZodObject`. Add a union member that says "not only accept ZodObject, but ALSO accept ZodEffect\<ZodObject\>"
2. When looking up the schema shape, do a very efficient check to see what type we are, and get the schema shape based on whether we are a ZodEffect or ZodObject

Add a test. Blamo. Easy!

There's a lot of discussion of "problems with z.discriminatedUnion". I see this as _independent_ from those critiques. This, I perceive, was just a basic oversight in the typing of the original implementation. The fix is simple and carries minimal overhead.

### CODE REVIEW

File: deno/lib/README.md [hunk 1]
```diff
@@ -57,8 +57,8 @@
 - [Coercion for primitives](#coercion-for-primitives)
 - [Literals](#literals)
 - [Strings](#strings)
-  - [Datetime](#datetime-validation)
-  - [IP](#ip-address-validation)
+  - [Datetime](#iso-datetimes)
+  - [IP](#ip-addresses)
```
Comment (cdaringe) [inline deno/lib/README.md:61]: uh errr uhhh i didn't do this. this was the git hooks? should i purge this?

File: deno/lib/types.ts [hunk 2]
```diff
@@ -2882,13 +2882,20 @@ const getDiscriminator = <T extends ZodTypeAny>(
   }
 };
 
-export type ZodDiscriminatedUnionOption<Discriminator extends string> =
```
Comment (cdaringe) [inline deno/lib/types.ts:2885]: this is a bad diff 😆 . this type is still present on L2895

File: deno/lib/types.ts [hunk 3]
```diff
@@ -2882,13 +2882,20 @@ const getDiscriminator = <T extends ZodTypeAny>(
...
     UnknownKeysParam,
     ZodTypeAny
   >;
 
+type ZodDiscriminatedUnionOptionEffectsObject<Discriminator extends string> =
+  ZodEffects<ZodDiscriminatedUnionOptionObject<Discriminator>>;
+
+export type ZodDiscriminatedUnionOption<Discriminator extends string> =
+  | ZodDiscriminatedUnionOptionObject<Discriminator>
+  | ZodDiscriminatedUnionOptionEffectsObject<Discriminator>;
```
Comment (cdaringe) [inline deno/lib/types.ts:2897]: the creation of this union is the **core change** in this PR!

**Comment by cdaringe** [comment]
hey friends, this is a pretty simple PR to review. let me know if there's anything i can do to reduce your review burden. i annotated the changes, for example, as a mechanism to make your review more straight forward. i know maintenance is a huge effort, so I want to be courteous of your precious time!

**Comment by cdaringe** [comment]
Hey all, friendly bump. Thanks for your time and support

**Comment by cdaringe** [comment]
Hey all, friendly bump. Thanks for your time and support

**Comment by ebramanti** [comment]
I have run into this issue too, would be really helpful if this change could land soon!

**Comment by k0rnl** [comment]
I have run into this issue too, would be really helpful if this change could land soon ! 🙏

**Comment by colinhacks** [comment]
This is a well-made PR, but the root problem here is that refinements should be stored *inside* the schema instead of wrapping the schema with ZodEffects. I'm hesitant to merge because this adds more complexity to a feature (discriminated union) that likely won't exist in Zod 4. The Zod 4 implementation will avoid the fundamental messiness of discriminatedUnion and it's right around the corner, so I'm gonna close. Thanks @cdaringe, this was a great PR.

**Comment by ms-ati** [comment]
Almost a year later, checking in?

**Pull request closed without merging**